core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -888,21 +873,14 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
                     mapper = context.docMapper().objectMappers().get(context.path().pathAsText(paths[i]));
                     if (mapper == null) {
                         // One mapping is missing, check if we are allowed to create a dynamic one.
-                        ObjectMapper.Dynamic dynamic = parent.dynamic();
-                        if (dynamic == null) {
-                            dynamic = dynamicOrDefault(context.root().dynamic());
-                        }
+                        ObjectMapper.Dynamic dynamic = dynamicOrDefault(parent, context.root().dynamic());
 
                         switch (dynamic) {
                             case STRICT:
                                 throw new StrictDynamicMappingException(parent.fullPath(), paths[i]);
                             case TRUE:
                                 Mapper.Builder builder = context.root().findTemplateBuilder(context, paths[i], "object");
                                 if (builder == null) {
-                                    // if this is a non root object, then explicitly set the dynamic behavior if set
-                                    if (!(parent instanceof RootObjectMapper) && parent.dynamic() != ObjectMapper.Defaults.DYNAMIC) {
-                                        ((ObjectMapper.Builder) builder).dynamic(parent.dynamic());
-                                    }
```

Comment:
I think what you are doing in `dynamicOrDefault` is different from what was done here. If you have a single level of objects this will be equivalent. But if you have several levels, then this logic used to propagate the value of `dynamic` down the tree while with what you are deing in `dynamicOrDefault` now, we would only look at the parent mapper?

we almost need to maintain a stack of the values of the `dynamic` property for all parent object mappers?